web/modules/custom/joinup_invite/src/Controller/InvitationController.php

```diff
@@ -0,0 +1,133 @@
+ *
+ * @see \Drupal\joinup_invite\Form\InviteToDiscussionForm
+ */
+class InvitationController extends ControllerBase {
+
+  /**
+   * The identifier for accepting an invitation.
+   *
+   * @var string
+   */
+  const ACTION_ACCEPT = 'accept';
+
+  /**
+   * The identifier for rejecting an invitation.
+   *
+   * @var string
+   */
+  const ACTION_REJECT = 'reject';
+
+  /**
+   * The helper service for creating and retrieving messages for invitations.
+   *
+   * @var \Drupal\joinup_invite\InvitationMessageHelperInterface
+   */
+  protected $invitationMessageHelper;
+
+  /**
+   * Constructs an InvitationController object.
+   *
+   * @param \Drupal\joinup_invite\InvitationMessageHelperInterface $invitationMessageHelper
+   *   The helper service for creating and retrieving messages for invitations.
+   */
+  public function __construct(InvitationMessageHelperInterface $invitationMessageHelper) {
+    $this->invitationMessageHelper = $invitationMessageHelper;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('joinup_invite.invitation_message_helper')
+    );
+  }
+
+  /**
+   * Accepts or rejects an invitation.
+   *
+   * @param \Drupal\joinup_invite\Entity\InvitationInterface $invitation
+   *   The invitation which is being accepted or rejected.
+   * @param string $action
+   *   The action which is being taken. Can be either 'accept' or 'reject'.
+   * @param string $hash
+   *   The hash value to protect against brute forcing invitations.
+   *
+   * @return \Symfony\Component\HttpFoundation\RedirectResponse
+   *   The redirect response.
+   */
+  public function updateInvitation(InvitationInterface $invitation, string $action, string $hash) : RedirectResponse {
+    $action === self::ACTION_ACCEPT ? $invitation->accept() : $invitation->reject();
+    $invitation->save();
+
+    $url = $invitation->getEntity()->toUrl();
+    return $this->redirect($url->getRouteName(), $url->getRouteParameters());
+  }
+
+  /**
+   * Access check for the invitation route.
+   *
+   * Access is granted only if the hash value is correct. This allows users to
+   * accept the invitation even if they are not currently logged in to the
+   * website.
+   *
+   * @param \Drupal\joinup_invite\Entity\InvitationInterface $invitation
+   *   The invitation which is being accepted or rejected.
+   * @param string $action
+   *   The action which is being taken. Can be either 'accept' or 'reject'.
+   * @param string $hash
+   *   The hash value to check.
+   *
+   * @return \Drupal\Core\Access\AccessResultInterface
+   *   The access result.
+   */
+  public function access(InvitationInterface $invitation, string $action, string $hash) : AccessResultInterface {
+    $valid_action = in_array($action, [self::ACTION_ACCEPT, self::ACTION_REJECT]);
+    return AccessResult::allowedIf($valid_action && static::generateHash($invitation, $action) === $hash);
```

Comment:
Why not? Why would you reject an invitation for this reason? It shouldn't matter who is logged in the website since the user is interacting with their mailbox. The access control should be handled on the mail client side, not on our side. If the hash is valid we know that the request is coming from the correct user.

There are various use cases where the user logged into joinup might be different from the user logged into the mail client:
* People might be working on a shared computer or tablet where some other user is logged in.
* A user might have multiple accounts on Joinup and is logged in with a different account.
* An assistant might be tasked with accepting an invitation on behalf of their superior.